shared/js/background.js

```diff
@@ -242,21 +246,44 @@ chrome.webRequest.onBeforeRequest.addListener(
             return
         }
 
+        if (thisTab.failedUpgradeUrls[requestData.url]) {
+            console.log('already tried https upgrades for this URL and failed, skip:\n' + requestData.url)
+            return
+        }
+
         // Avoid redirect loops
         if (thisTab.httpsRedirects[requestData.requestId] >= 7) {
             console.log('HTTPS: cancel https upgrade. redirect limit exceeded for url: \n' + requestData.url)
+
             return {redirectUrl: thisTab.downgradeHttpsUpgradeRequest(requestData)}
         }
```

Comment:
The `main_frame` HTTPS upgrades are addressed by the logic below, so this now only handles redirect loops for page assets (e.g. images, etc).

While we could port the `main_frame` time-based logic to prevent repeated requests for these, it seems a bit overkill.